File: Data Structures/HashSet/Python/HashSet.py
'''
 /$$   /$$                     /$$        /$$$$$$              /$$    
| $$  | $$                    | $$       /$$__  $$            | $$    
| $$  | $$  /$$$$$$   /$$$$$$$| $$$$$$$ | $$  \__/  /$$$$$$  /$$$$$$  
| $$$$$$$$ |____  $$ /$$_____/| $$__  $$|  $$$$$$  /$$__  $$|_  $$_/  
| $$__  $$  /$$$$$$$|  $$$$$$ | $$  \ $$ \____  $$| $$$$$$$$  | $$    
| $$  | $$ /$$__  $$ \____  $$| $$  | $$ /$$  \ $$| $$_____/  | $$ /$$
| $$  | $$|  $$$$$$$ /$$$$$$$/| $$  | $$|  $$$$$$/|  $$$$$$$  |  $$$$/
|__/  |__/ \_______/|_______/ |__/  |__/ \______/  \_______/   \___/ 
'''

'''
- EXPLANATION:
    - In Python, there is no built-in class named HashSet. 
      However, Python provides a built-in class called set, which can be used to achieve similar functionality to a hash set. 
      The set class in Python represents an unordered collection of unique elements. 
      It is implemented using a hash table data structure, which allows for efficient membership testing, insertion, and deletion operations.
      
      
- KEY FEATURES OF HASHSET:

    - Unique elements: Sets only contain unique elements. 
      If you try to add a duplicate element to a set, it will be ignored.
    
    - Unordered: The elements in a set are not stored in any particular order. 
      Therefore, you cannot access elements in a set by their indices.
    
    - Mutable: Sets are mutable, which means you can add and remove elements from them.
    
    - Hash-based: Sets in Python are implemented using a hash table, which allows for fast insertion, deletion, and membership testing.
    
    - Membership testing: You can easily check whether an element is present in a set using the in operator. 
      This operation has an average time complexity of O(1).
    
    - Set operations: Sets support various set operations such as union, intersection, difference, and symmetric difference. 
      These operations can be performed using built-in methods or operators.
      
      
- HERE ARE SOME OF THE MOST COMMONLY USED BUILT-IN FUNCTIONS:

    - add(element): Adds an element to the set.

    - remove(element): Removes an element from the set. Raises a KeyError if the element is not found.

    - discard(element): Removes an element from the set if it is present. Does not raise an error if the element is not found.

    - clear(): Removes all elements from the set, making it an empty set.

    - copy(): Returns a shallow copy of the set.

    - union(*sets): Returns a new set that is the union of the current set and the provided sets.

    - intersection(*sets): Returns a new set that is the intersection of the current set and the provided sets.

    - difference(*sets): Returns a new set that contains elements present in the current set but not in the provided sets.

    - symmetric_difference(other_set): Returns a new set that contains elements present in either the current set or the provided set, but not in both.

    - issubset(other_set): Checks if the current set is a subset of the provided set and returns a boolean value.

    - issuperset(other_set): Checks if the current set is a superset of the provided set and returns a boolean value.

    - isdisjoint(other_set): Checks if the current set and the provided set have no common elements and returns a boolean value.

    - len(set): Returns the number of elements in the set.

    - iter(set): Returns an iterator over the set, allowing you to iterate through its elements.
'''


# Creating a set
my_set = set()

# Adding elements to the set
my_set.add(1)
my_set.add(2)
my_set.add(3)
my_set.add(1)  # Ignored, as 1 is already present

print(my_set)  # Output: {1, 2, 3}

# Removing an element from the set
my_set.remove(2)

print(my_set)  # Output: {1, 3}

# Checking membership
print(1 in my_set)  # Output: True
print(2 in my_set)  # Output: False

# Set operations
set1 = {1, 2, 3}
set2 = {2, 3, 4}

union_set = set1.union(set2)
print(union_set)  # Output: {1, 2, 3, 4}

intersection_set = set1.intersection(set2)
print(intersection_set)  # Output: {2, 3}

difference_set = set1.difference(set2)
print(difference_set)  # Output: {1}

symmetric_difference_set = set1.symmetric_difference(set2)
print(symmetric_difference_set)  # Output: {1, 4}

File: Data Structures/HashSet/JavaScript/HashSet.js
/*
 /$$   /$$                     /$$        /$$$$$$              /$$    
| $$  | $$                    | $$       /$$__  $$            | $$    
| $$  | $$  /$$$$$$   /$$$$$$$| $$$$$$$ | $$  \__/  /$$$$$$  /$$$$$$  
| $$$$$$$$ |____  $$ /$$_____/| $$__  $$|  $$$$$$  /$$__  $$|_  $$_/  
| $$__  $$  /$$$$$$$|  $$$$$$ | $$  \ $$ \____  $$| $$$$$$$$  | $$    
| $$  | $$ /$$__  $$ \____  $$| $$  | $$ /$$  \ $$| $$_____/  | $$ /$$
| $$  | $$|  $$$$$$$ /$$$$$$$/| $$  | $$|  $$$$$$/|  $$$$$$$  |  $$$$/
|__/  |__/ \_______/|_______/ |__/  |__/ \______/  \_______/   \___/ 
*/

/*
- EXPLANATION:

    - A HashSet is a data structure that represents a collection of unique elements, where each element appears only once in the collection. 
      It is designed to efficiently store and manage a set of distinct values without allowing duplicates. 
      HashSets typically provide fast insertion, deletion, and lookup operations, 
      making them useful for scenarios where you need to quickly determine the presence of an element or 
      ensure that no duplicates exist within the collection.


- KEY FEATURES:

      Uniqueness: The primary feature of a HashSet is that it stores unique elements, meaning that duplicate values are not allowed. 
      Each element can appear only once in the set.

      Fast Lookup: HashSets use a hashing function to map elements to specific locations in memory, allowing for efficient retrieval of elements. 
      On average, lookup operations have constant time complexity O(1), making it ideal for quick membership checks.

      Fast Insertion and Deletion: Similarly, insertion and deletion operations in a HashSet are typically fast with constant time complexity O(1) on average. 
      The hashing function helps determine the exact position of the element, simplifying these operations.

      No Guaranteed Order: HashSets do not maintain any specific order of elements. 
      The order in which elements are inserted may not be the same as the order in which they are iterated.

      Dynamic Size: HashSets can dynamically grow or shrink as elements are added or removed. 
      They automatically manage memory allocation and resizing to accommodate changes in the number of elements.

      Common Set Operations: HashSet supports standard set operations like union, intersection, difference, and subset checks, 
      making it useful in various algorithmic scenarios.

      Common Use Cases: HashSets are commonly used in scenarios where unique elements need to be efficiently managed, 
      such as removing duplicates from a list, checking for the existence of elements, or storing a collection of unique items.

      Implementation Variations: Depending on the programming language, HashSet implementations may differ. 
      Some languages provide built-in HashSet data structures (e.g., Java, C#), 
      while others use standard libraries or simulate HashSet behavior using objects or other data structures (e.g., JavaScript).

      Hashing and Collision Handling: The efficiency of HashSets depends on the quality of the hashing function used. 
      In cases of hash collisions (where two different elements are mapped to the same location), 
      various collision resolution techniques are employed to handle and manage these situations.

      Memory Overhead: HashSets generally have a slightly higher memory overhead compared to simple arrays or 
      lists due to the need for maintaining the underlying hash table and managing collision resolution.
*/

/*
- LEETCODE EXERCISES:
    - Two Sum (Easy): https://leetcode.com/problems/two-sum/
    - Contains Duplicate (Easy): https://leetcode.com/problems/contains-duplicate/
    - Happy Number (Easy): https://leetcode.com/problems/happy-number/
    - Isomorphic Strings (Easy): https://leetcode.com/problems/isomorphic-strings/
    - Valid Sudoku (Medium): https://leetcode.com/problems/valid-sudoku/
    - Longest Consecutive Sequence (Hard): https://leetcode.com/problems/longest-consecutive-sequence/
*/
